### PR TITLE
Update Tripal Importer database rollback to apply to Chado, not the Drupal database

### LINKED ISSUE #1764: Tripal Importer class doesn't appear to be rolling back database transactions when exception is thrown
### Tripal Version

4.0.alpha3

### Drupal Version

10.0.12-dev

### PostgreSQL Version

13.13

### PHP Version

 8.1.27

### General Description

I've been developing a new importer and Lacey and I have been having trouble getting the database transaction to roll back once an exception has been thrown: https://github.com/TripalCultivate/TripalCultivate-Germplasm/pull/6#pullrequestreview-1778536959 

I decided to test the FASTA importer to see if I could reproduce the problem there. I did this by adding an if statement in the FASTA importer code to throw an exception after the 5th feature in my FASTA file. Once I ran the job, my error was successfully printed to the terminal, but the first 5 features were committed to the database (out of the 9 total features in the file).

### Steps to reproduce

1. I cloned the tripal repo and made sure I was on the `4.x` branch, then created a fresh tripaldocker container that mounted to this repo so I could make code changes:
```
docker run --publish=8880:80 --name=t4 -tid --volume=$(pwd):/var/www/drupal/web/modules/contrib/tripal tripalproject/tripaldocker:latest
docker exec t4 service postgresql start
```
2. I added the following code to line 645 of `tripal_chado/src/Plugin/TripalImporter/FASTAImporter.php`:
```
      if ($j == 5) {
        throw new \Exception(
          t("Reached sequence 5, exception thrown.")
        );
      }
```
3. I logged into my new tripal site at `http://localhost:8880` and created an Organism (**Citrus sinensis**) and Analysis (**Whole Genome Assembly and Annotation of Citrus sinensis**) through the interface

4. For my test FASTA file ([Citrus sinensis-orange1.1g015632m.g.fasta](http://tripal.info/sites/default/files/Citrus_sinensis-orange1.1g015632m.g.fasta)), I grabbed it from the Tripal 3 docs here: https://tripal.readthedocs.io/en/latest/user_guide/example_genomics/genomes_genes.html 

5. I filled out the FASTA Importer form as follows:

Field Name | Value
-- | --
FASTA file | Upload the file named Citrus_sinensis-orange1.1g015632m.g.fasta
Analysis | Whole Genome Assembly and Annotation of Citrus sinensis
Organism | Citrus sinensis (Sweet orange)
Sequence type | mRNA
Method | Insert and update*
Name Match | Name

*Note that this is the only change I made from the Tripal 3 docs since there are currently no features loaded in this site

6. I submitted the form and ran my job on the command line:
```
docker exec t4 drush trp-run-jobs --job_id=1 --username=drupaladmin --root=/var/www/drupal/web
```
See output under "Error Messages".

7. I checked the database using `docker exec -it t4 drush php:cli` to see that the first 5 sequences were in the chado.feature table:
```
$connection = \Drupal::service('tripal_chado.database');
= Drupal\tripal_chado\Database\ChadoConnection {#6115}

> $feature_query = $connection->select('1:feature','f')->fields('f', ['organism_id', 'name', 'uniquename', 'type_id']);
= Drupal\pgsql\Driver\Database\pgsql\Select {#6176}

> $feature_record = $feature_query->execute()->fetchAll();
= [
    {#6141
      +"organism_id": "1",
      +"name": "orange1.1g015632m",
      +"uniquename": "orange1.1g015632m",
      +"type_id": "166",
    },
    {#6140
      +"organism_id": "1",
      +"name": "orange1.1g015662m",
      +"uniquename": "orange1.1g015662m",
      +"type_id": "166",
    },
    {#6130
      +"organism_id": "1",
      +"name": "orange1.1g018514m",
      +"uniquename": "orange1.1g018514m",
      +"type_id": "166",
    },
    {#6138
      +"organism_id": "1",
      +"name": "orange1.1g015615m",
      +"uniquename": "orange1.1g015615m",
      +"type_id": "166",
    },
    {#6137
      +"organism_id": "1",
      +"name": "orange1.1g015645m",
      +"uniquename": "orange1.1g015645m",
      +"type_id": "166",
    },
  ]
```

### Error messages

```shell
2024-01-24 22:01:38
Tripal Job Launcher
Running as user 'drupaladmin'
-------------------
2024-01-24 22:01:38: Job ID 1.
2024-01-24 22:01:38: Calling: tripal_run_importer(5)
Running 'Chado FASTA File Loader' importer
 [notice] Running 'Chado FASTA File Loader' importer
NOTE: Loading of this file is performed using a database transaction. If it fails or is terminated prematurely then all insertions and updates are rolled back and will not be found in the database
 [notice] NOTE: Loading of this file is performed using a database transaction. If it fails or is terminated prematurely then all insertions and updates are rolled back and will not be found in the database
Step 1: Finding sequences...
 [notice] Step 1: Finding sequences...
Step 2: Importing sequences...
 [notice] Step 2: Importing sequences...
Found 9 sequence(s).
 [notice] Found 9 sequence(s).
Percent complete: 0%. Memory: 32,514,184 bytes.
 [notice] Percent complete: 0%. Memory: 32,514,184 bytes.
Percent complete: 0%. Memory: 32,661,680 bytes.
 [notice] Percent complete: 0%. Memory: 32,661,680 bytes.
Percent complete: 11.11 %. Memory: 32,661,680 bytes.
 [notice] Percent complete: 11.11 %. Memory: 32,661,680 bytes.
Percent complete: 22.22 %. Memory: 32,661,784 bytes.
 [notice] Percent complete: 22.22 %. Memory: 32,661,784 bytes.
Percent complete: 33.33 %. Memory: 32,661,912 bytes.
 [notice] Percent complete: 33.33 %. Memory: 32,661,912 bytes.
Percent complete: 44.44 %. Memory: 32,661,912 bytes.
 [notice] Percent complete: 44.44 %. Memory: 32,661,912 bytes.
ERROR: Reached sequence 5, exception thrown.
 [error]  ERROR: Reached sequence 5, exception thrown. 
[site http://default] [TRIPAL] ERROR: Reached sequence 5, exception thrown.
```

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by carolyncaron**
# Bug Fix

## Description

Tripal importers are designed to roll back a database transaction if an error occurs during the import, but as found in issue #1764, there was a bug preventing the rollback. It turns out the rollback was occurring if anything was inserted into the Drupal database, but completely ignored Chado (possibly a relic of how this was done in Drupal 7). After discussions with @laceysanderson, we determined that the rollback should not occur for the Drupal database at all, since log messages are stored there which the user will likely want accessible to troubleshoot their failing import job. Rollback in a different database  (such as Drupal) should instead be handled by individual importers. This PR updates the rollback to occur by default in Chado instead.

### Manual testing
You can follow the steps to reproduce the bug in issue #1764, and then switch to this branch and repeat the process again, this time confirming that Chado has no new records. 

### Automated tests
I created a new kernel test class to override the base importer run method and simulate the following situations:

- [x] **testTransactionRollback()**: Inserts into the organism table in Chado, then throws an exception.
   - Asserts an exception is in fact thrown
   - Asserts chado.organism does not have anything added
- [x] **testTransactionCommit()**: Inserts into the organism table in Chado, no error is thrown
   - Asserts that an exception does not get thrown
   - Asserts that chado.organism has a new organism added since no errors occurred

### CODE REVIEW

**Review by laceysanderson — COMMENTED** [review]
Thanks for submitting the PR @carolyncaron and for tracking down the issue 😃! The current fix does add a circular dependency (i.e. it refers to chado from the tripal/tripal module and thus will fail if tripal_chado is not enabled).

I think a good way to fix this is to add a method to the importer classes that allows them to say how to start a transaction. This way we can move the code you added into the ChadoImporterBase (i.e. withing tripal_chado) but it will still be available in tripal/tripal for chado-based importers only.

**Review by laceysanderson — COMMENTED** [review]
I added in my changes and expanded the test to also confirm it ONLY rolls back chado and not Drupal. I fully approve this PR but now that it has some of my code/logic in it, **it may be good for someone else from the Tripal team to give it a test or even just a look over**.

File: tripal_chado/tests/src/Kernel/Plugin/TripalImporter/ChadoImporterTransactionTest.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    // Ensure we see all logging in tests.
+    \Drupal::state()->set('is_a_test_environment', TRUE);
+
+		// Open connection to Chado
+		$this->connection = $this->getTestSchema(ChadoTestKernelBase::PREPARE_TEST_CHADO);
```
Comment (dsenalik) [inline tripal_chado/tests/src/Kernel/Plugin/TripalImporter/ChadoImporterTransactionTest.php]:
```suggestion
    // Open connection to Chado
    $this->connection = $this->getTestSchema(ChadoTestKernelBase::PREPARE_TEST_CHADO);
```

File: tripal_chado/tests/src/Kernel/Plugin/TripalImporter/ChadoImporterTransactionTest.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    'file_upload' => FALSE,
+    'file_load' => FALSE,
+    'file_remote' => FALSE,
+    'file_required' => FALSE,
+    'cardinality' => 1,
+  ];
+
+  protected $importer;
+
+	/**
```
Comment (dsenalik) [inline tripal_chado/tests/src/Kernel/Plugin/TripalImporter/ChadoImporterTransactionTest.php]:
```suggestion
  /**
```

**Review by dsenalik — APPROVED** [review]
I followed the test procedure with the code to throw an exception at 5, and confirmed no features were stored in chado.
I removed the code to throw exception at 5, and ran again, and nine features were stored.
From code review I have a couple of suggested whitespace changes (tab characters) but since this these are not functional changes, I will go ahead and mark as approved.

**Comment by laceysanderson** [comment]
Thank you @dsenalik! ❤️ 🎉  I added in your whitespace changes so once tests pass again we can merge! 😍

**Pull request merged**